### ISSUE TITLE
Is there interest for an excercism GUI client?

### ISSUE THREAD

**Issue by holandes22** [exercism/exercism.io#2939]
Hi,

I recently heard the bikeshed podcast with @kytrinyx (http://bikeshed.fm/65), and near the end she mentions that would be nice to have a GUI client for users that are not familiar with the command line or don't know what a PATH envar is.

I've been tinkering the last few days with [electron](http://electron.atom.io/), and seems that it might be a viable solution for a cross-platform GUI client. 
The thing is,  I don't have too much bandwidth and, even tough is an interesting project, I'd like to invest in something that can be useful to some. So, I'm asking here the community: _is  there interest in having a GUI client, as a companion for the CLI?_.

I'm familiar with ember.js so I used that to build the demo app along with the excelent [ember-electron](https://github.com/felixrieseberg/ember-electron) addon that helps integrate it with electron. 

Here is the source code https://github.com/holandes22/exercism-gui and here is a demo video:

<a href="http://www.youtube.com/watch?feature=player_embedded&v=kLgFGxbv68M
" target="_blank"><img src="http://img.youtube.com/vi/kLgFGxbv68M/0.jpg" alt="demo" width="240" height="180" border="10" /></a>

The idea is to have feature parity with the CLI client for a first release. For the demo I just implemented:

- Read and write config file
- List and detail tracks
- Show track status

So we are missing:

- Fetch
- Submit
- Skip
- Restore
- Open (show link to latest iteration o some problem)

It should be multi-platform (Linux, Windows and MacOS) and should be able to co-exist with the CLI. 

## Packaging

Here is a video showing how to make a package. 

<a href="http://www.youtube.com/watch?feature=player_embedded&v=fu_XvHxHuXs
" target="_blank"><img src="http://img.youtube.com/vi/fu_XvHxHuXs/0.jpg" alt="packaging" width="240" height="180" border="10" /></a>

I'm building for Linux in this case as that is the platform I have, compiling for other platforms requires those platforms as build machines or using [Wine (https://www.winehq.org/))

## TODO list if moving forward

- Opend GH tickets for the tasks and features to implement
- Add a contributing guide in readme (setting dev env, tests, building, etc)
- Set CI for automatic test running and package building
- Add tests
- Clean up code from the demo before moving forward with other tasks.
- UI design

## Remarks / Challenges

Disclaimer: I have no experience with electron or node.js so some items here come from lack of knowledge in the area, so this might be trivial to someone more experienced.

- Packaging for different platforms. ember-electron allows you to build for all platforms, but to build from my Linux rig other platforms I need to mess around with Wine (which probably does not cover MacOS case). Need to find a solution that allows to build for all platforms, and more importantly that works with a CI system to automate releases. 
- Testing MacOS and Windows platforms. I have no or very limited access to those platforms, so testing requires some help from people who do.
- Making installers. Electron apps are effectively desktop apps, so it is also possible to have installers/distro packages. This might not be needed as we can simple distribute a compressed folder containing the electron package, but it would definitely be nice to have. Here I have no experience whatsoever on Mac or Windows installers. For Linux, having packages for the main distros (Ubuntu, Fedora, Arch, etc.) might be enough.
- SSL: The backend is currently not using SSL and having the client connecting to an unverified server is quite scary when dealing with desktop apps.

__Sorry for the looooooong issue!!!__

**Comment by kytrinyx** [comment]
> Sorry for the looooooong issue!!!

Holy smokes—this is so cool!

OK, my first impression is that this would be super valuable to people who aren't comfortable with the command-line, and also that it's likely to be a huge amount of work. I personally can't devote resources to it, since I'm drowning under all the existing things I *should* be doing.

I'm talking to people about the SSL thing; Heroku has made some things available that would make SSL affordable for us.

I don't know how to make installers. I think that technically we could write installers for the CLI, which would help alleviate some of the pain that people have when they don't understand what PATH is.

I'm pretty sure we could get people to help test on other platforms.

I really love how this would lower the barrier to using the site, and I wouldn't want to have something where a single person is single-handedly working on such a huge and important feature (the risk is really high for burning out—even one second person tends to make this a much less scary proposition).

I've been wanting to rethink pretty much every aspect of the user experience, and this issue fits right in with that. Here's the issue I posted about rethinking the command-line: https://github.com/exercism/discussions/issues/35

I'll mention this issue in the weekly-ish "behind the scenes" newsletter, and also tweet about it to see if there is anyone else who would be excited to explore this.

**Comment by holandes22** [comment]
@kytrinyx awesome, thank you! This encourages me to keep working on the project :)

So I'll work on cleaning up the project, move on with the features, CI and looking for options to build for Windows/Mac. Hope I get some traction and that attracts some contributors to help out with the things I have no experience or lack resources (Windows/Mac)

Been thinking a little more on the lack of SSL and the potential abuses of the client and it quite concerns me (concerns that also affects the CLI), so I think these should be addressed. I can DM you with more specifics on this.

**Comment by kytrinyx** [comment]
Yeah, I agree that the SSL needs to be addressed. Over the next six weeks I will not be able to tackle it, unless someone else takes point on figuring it out. I'll put the word out and see if someone has the capacity to run with it before I do.

**Comment by holandes22** [comment]
Just two things I'm not sure about:

1. Is there any problem in using the image assets from this repo, from a licensing perspective? (I used the site logo and the logos for the tracks) 
2. I think we should leave this ticket open for a little while to allow people to comment on it, but please feel free to close it when you think is appropriate

**Comment by ErikSchierboom** [comment]
I love the suggestion, and I might even be able to help. Personally, I think a GUI version would lower the barrier for a lot of people, which can only be good, right?

**Comment by mondayrain** [comment]
This looks amazing. 

I've been meaning to jump into learning electron.io, so although I would be slow help, I'd definitely be willing. I also have an OSX for the summer, so could help with testing installation at least up until September rolls around. I think a binary installer would be easiest for those who don't even know what the command line is... on the other hand, people probably need to be introduced to it at some point. The question is when, and I'm just not sure _right at the beginning_ is the best time.

In any case, if this gets the go-ahead, I'd 100% love to help!

**Comment by kytrinyx** [comment]
> I'm just not sure right at the beginning is the best time

Yeah, I completely agree. Get people started first, then introduce the command-line once they're comfortable with how the site works.

**Comment by kytrinyx** [comment]
> Is there any problem in using the image assets from this repo, from a licensing perspective? (I used the site logo and the logos for the tracks)

There shouldn't be, so long as the proper attribution is given for the track icons (we have an ATTRIBUTION.md file)

**Comment by holandes22** [comment]
@kytrinyx thank you, I'll use that attribution file to give proper credit.

I finally got some time today to update the README and open tasks (using GH issues).

@mondayrain @ErikSchierboom thanks for the feedback! I'll be using waffle.io to track ticket progress, you can check it out here https://waffle.io/holandes22/exercism-gui and see if something looks like fun to do. 

Priorities are:
- Having a proper guide for contributors
- having packages built automatically so people can start testing them
- Sort out the best way to write and run tests
- Clean up the demo code

That is the plan for this week at least

**Comment by mondayrain** [comment]
Looked at the Waffle, looks fun! Anything on the priority list that you'd
like to delegate, or do you want us jumping straight to features? I took a
look at some of the priority issues you made and it looks like you've got a
good start/handle on it.

**Comment by holandes22** [comment]
Yes, currently working on getting CI to build the packages. 
I really must clean up the demo code and have some tests setup as example for contributors, so I think I'll move on to that and the come back to packaging.

I think https://github.com/holandes22/exercism-gui/issues/14 can be a good ticket to start, easy to do with services (if you are familiar with that ember concept https://guides.emberjs.com/v2.6.0/applications/services/). You can also check the ones marked as feature and see what you find interesting

BTW, Is there an exercism gitter/irc/slack channel we can use to communicate?

**Comment by mondayrain** [comment]
On it. I need to learn Ember + a bit of Electron before I can really get
started but I'll try to have a PR to you by end of the week! Stoked about
this!

**Comment by holandes22** [comment]
sure, no rush, try to get to know ember, is really cool in my opinion!

I should have properly read the README, I saw the gitter link. I'll hang in the dev room there if you have any questions

**Comment by holandes22** [comment]
So I'm closing this as it did it's purpose. You can check progress of the project at 
https://waffle.io/holandes22/exercism-gui

Got CI runnnig tests and building packages for all platforms, so next step is to focus on developing features and polishing for a first alpha release.

You can ping me on the gitter channel if you have questions or open an issue at the project repo

**Comment by kytrinyx** [comment]
Excellent! @holandes22 let me know if you want to move the GUI to the exercism organization. I'll make a team for it, and we can add whoever you want to it as collaborators.

**Comment by holandes22** [comment]
@kytrinyx moving the project under the exercism organization is a great idea! Just let me know what I need to do from my part to migrate/transfer the repo

**Comment by kytrinyx** [comment]
I've created a team for the GUI—you should have an invitation. Accept that. If you go to the settings you should find a section at the bottom called "danger zone". One option there is to transfer the repository. Click that and follow the instructions.

Once you've done that let me know and I'll add the repository to the team with the correct permissions. This will allow you to push your changes up to it.

There are redirects in place by default so you technically wouldn't have to do anything else, but I tend to also repoint the origin:

```
git remote rm origin
git remote add origin $PATH_TO_NEW_GIT_REMOTE
```

Let me know if you have any questions.

**Comment by holandes22** [comment]
@kytrinyx thank you, got the invitation.

I'll try to get the migration done during the weekend. BTW, do you by any chance know if travis and appveyor handle repo migrations automatically or there is some work to do to in that space?

**Comment by kytrinyx** [comment]
I think I will have to turn something on for travis. I've never used appveyor so I don't know about that one.

**Comment by holandes22** [comment]
@kytrinyx I accepted the invite to the organization and then tried to transfer the repo to the exercism org, but got a permissions error:

![transfer](https://cloud.githubusercontent.com/assets/510233/16337042/186c19fc-3a1b-11e6-8e56-82b5df473a93.png)

According to this https://help.github.com/articles/transferring-a-repository-owned-by-your-personal-account/ if I transfer to an org and, I belong to that org, then I need to have admin/owner permissions. Maybe I'll need to leave the org, transfer, and then rejoin.

**Comment by kytrinyx** [comment]
> I accepted the invite to the organization and then tried to transfer the
repo to the exercism org, but got a permissions error:

I'll upgrade your access tonight when I get proper internet access.

**Comment by kytrinyx** [comment]
@holandes22 Would you transfer the repository to my user account (`kytrinyx`)? That will let me transfer it to the exercism organization.

**Comment by holandes22** [comment]
Sure thing. Will get to it tomorrow (my evening probably)

**Comment by holandes22** [comment]
@kytrinyx sent you a transfer request to your account. Please let me know once the transfer is complete so I can verify tests and builds continue to run in CI (travis and appveyor)

**Comment by kytrinyx** [comment]
Excellent!

**Comment by kytrinyx** [comment]
I didn't get the transfer request—would you double-check this?

**Comment by kytrinyx** [comment]
Oh, hah. It came to a different email address. That was unexpected. Give me a sec.

**Comment by kytrinyx** [comment]
Ok, done. It's now at exercism/gui and you have admin access to the repo, @holandes22

**Comment by holandes22** [comment]
Awesome, thank you!

**Comment by holandes22** [comment]
@kytrinyx hi,                                                                                                                                                      
I've been making some good progress in the GUI and think is ready for a first Alpha. Feature-wise is on pair with the CLI, except for `download` (I'll probably get to it next week). Latest release is here https://github.com/exercism/gui/releases/latest                                                                                                           
                                                                                                                                                                                 
I have a few questions and maybe you can help me with them:
                                                                                                                                                                        
- I've been having trouble spreading the word about the project and getting contributions, do you think you can mention it in the newsletter? All help is welcome, but basically what I need the most is help testing on Windows/Mac and feedback (if things are unclear, overly complicated, slow, missing features, bad UX/UI, etc)
                                                     
- (This will be relevant after the first final release) I'm a little concerned that the GUI might be a source of confusion. The GUI is not intended to be a replacement for the site, but to give a graphical interface to the API. 
A new user might not quite understand which things should be managed from the GUI (fetching, submitting, etc) and which ones should be managed on the site (reviewing/manage submissions, etc). Any thoughts on how we might explain this?                                                                                                                                        

I should mention that I didn't have time to tackle the installer/auto-update yet (from a first look seems complicated and quite a lot of work, someone with experience in electron that can help me here would be great) so for now, `installation` is just manually opening the tar.gz/zip package.                                                                     
                                                                                                                                                                                 
Also, please let me know if you prefer a different communication channel (gitter or other GH issue) 

Thanks!

**Comment by kytrinyx** [comment]
This is so exciting!

> I've been having trouble spreading the word about the project and getting contributions, do you think you can mention it in the newsletter?

Yes! I've missed the newsletter for the past two weeks (I was preparing to speak at GopherCon in front of 1500 people, and everything else kind of got dropped by the wayside). But yes: this Sunday I'll mention it. And also give it some attention on twitter.

> Any thoughts on how we might explain this?

The GUI would be instead of the CLI, right? So if we can figure out how to explain how Exercism works in general, then we should be in a pretty good place. E.g.: write code in your local env vs browse and have conversations about code on exercism.io, use either the CLI or the GUI to bridge the gap between the two (fetch, submit), @exercism/rgsoc2016 what do you think?

> didn't have time to tackle the installer/auto-update yet

No worries! Just getting a working GUI is super exciting, and we can figure out details like this in a second pass.

> Also, please let me know if you prefer a different communication channel (gitter or other GH issue)

I don't have strong preferences. At work we try to open different issues for separate requests/discussions, so if that sounds good we could use new issues in the exercism/gui repository—just remember to ping me with an `@`-mention, otherwise it will take longer before I see it.

**Comment by holandes22** [comment]
@kytrinyx thank you! I think a mention in the newsletter/twitter will really help spreading the word. 
I won't be available during the last 3 weeks of August, hence I aim to have the first beta out by the beginning of that month, so getting feedback/issues reports during the next few weeks would be great.
                                                                                                                                             
Thank you for your input on explaining the roles of gui/cli/site, gives me an starting point. My plan is,  after we reach a stable point (I'm guessing mid September) to open a PR to exercism.io adding details/help for the GUI as an alternative to the CLI and we'll iterate on that.                           

BTW 
> The GUI would be instead of the CLI, right? 

I hope not, just an alternative. I actually prefer the CLI myself :D 

                                                                                                                                                                             
As for communication channel, I think that GH issues works best too. I'll open issues in exercism/gui and @-mention you if it requires your input specifically 

Thanks again!

**Comment by kytrinyx** [comment]
> I hope not, just an alternative. I actually prefer the CLI myself :D

Oops, that's what I meant :) Someone might use the GUI instead of the CLI (but I think a lot of people are happy on the command-line).

Super excited about this!
